src/strtree.c

```diff
@@ -91,20 +99,31 @@ static PyObject *STRtree_new(PyTypeObject *type, PyObject *args,
     }
 
     n = PyArray_SIZE((PyArrayObject *) arr);
+
+    kv_init(_geoms);
     for(i = 0; i < n; i++) {
         /* get the geometry */
         ptr = PyArray_GETPTR1((PyArrayObject *) arr, i);
         obj = *(GeometryObject **) ptr;
         /* fail and cleanup incase obj was no geometry */
         if (!get_geom(obj, &geom)) {
             GEOSSTRtree_destroy_r(context, tree);
+            // free the geometries
+            count = kv_size(_geoms);
+            for (i = 0; i < count; i++) { Py_XDECREF(kv_pop(_geoms)); }
+            kv_destroy(_geoms);
             return NULL;
         }
         /* skip incase obj was None */
-        if (geom == NULL) { continue; }
+        if (geom == NULL) {
+            kv_push(GeometryObject *, _geoms, NULL);
+        } else {
         /* perform the insert */
-        count++;
-        GEOSSTRtree_insert_r(context, tree, geom, (void *) i );
+            Py_INCREF(obj);
+            kv_push(GeometryObject *, _geoms, obj);
```

Comment:
Would an alternative be to loop through `arr` and INCREF each element of it? (and keep it as a numpy array) Or is there a specific benefit of putting those in a separate vector?

Reply:
The array itself is mutable, so in the case the array is changed after constructing the tree you might end up with a memory leak or a segfault

Reply:
Ah, yes, good point